### PR TITLE
Replace 8 with real maxMipLevel in lights_fragment_maps.glsl

### LINKED ISSUE #13107: MeshStandardMaterial: Reflections visible for moderate roughness values
Comparing [MetalRoughSpheres](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/MetalRoughSpheres) in a few different engines:

| engine | screenshot |
|---|---|
| three.js | screen shot 2018-01-13 at 12 56 00 pm |
| babylon.js | screen shot 2018-01-13 at 12 56 08 pm |
| qtek | screen shot 2018-01-13 at 12 55 02 pm |

THREE.MeshStandardMaterial is noticeably more reflective at moderate roughness values compared to other engines, and perhaps smoother/shinier than content artists would intend.

@emackey brought this up on another thread, and mentioned that Blender and Substance Painter also look similar to the other engines.

### PR DESCRIPTION
**Pull request by takahirox**
This PR is still WIP but you can test if this change fixes #13107
I calculate max mip level by log2( max( image.width, image.height ) ) for gl.generateMipmap.

I'll add comments later about what I'm worried about in this impl.

### CODE REVIEW

File: src/renderers/shaders/ShaderChunk/lights_fragment_maps.glsl [hunk 1]
```diff
@@ -25,11 +25,10 @@
 
 #if defined( USE_ENVMAP ) && defined( RE_IndirectSpecular )
 
-	// TODO, replace 8 with the real maxMIPLevel
```
Comment (emackey) [inline src/renderers/shaders/ShaderChunk/lights_fragment_maps.glsl]: Just curious, this file appears to contain two copies of this `TODO` comment, but only one got removed.  Was that intentional?

Comment (takahirox) [reply src/renderers/shaders/ShaderChunk/lights_fragment_maps.glsl]: Ah, I'm just missing another one.
I'll update.

Comment (takahirox) [reply src/renderers/shaders/ShaderChunk/lights_fragment_maps.glsl]: Updated, thanks!

File: src/renderers/webgl/WebGLTextures.js [hunk 2]
```diff
@@ -81,6 +81,16 @@ function WebGLTextures( _gl, extensions, state, properties, capabilities, utils,
 
 	}
 
+	function generateMipmap( texture, target ) {
+
+		_gl.generateMipmap( target );
+
+		var image = Array.isArray( texture.image ) ? texture.image[ 0 ] : texture.image;
```
Comment (takahirox) [inline src/renderers/webgl/WebGLTextures.js]: Can I expect images for cubic map have the same size?

**Comment by takahirox** [comment]
Changes and notes.
- Using uniforms for maxMipLevel in glsl rather than using #define because texture can be updated.
- Calculating max mip level by log2( max( image.width, image.height ) ) for gl.generateMipmap.
- Holding maxMipLevel in textureProperties.

I add some questions inline.

File: src/renderers/webgl/WebGLTextures.js [hunk 3]
```diff
@@ -81,6 +81,16 @@ function WebGLTextures( _gl, extensions, state, properties, capabilities, utils,
...
+		var textureProperties = properties.get( texture );
+		textureProperties.__maxMipLevel = Math.max( Math.log2( Math.max( image.width, image.height ) ), textureProperties.__maxMipLevel );
```
Comment (takahirox) [inline src/renderers/webgl/WebGLTextures.js]: Will gl.generateMipmap() clear the existing mipmap configuration?
For example, imagine you have 64x64 texture image and set mip level up to 10 with gl.texImage2D and then call gl.generateMipmap. gl.generateMipmap sets mip levels 1-6. So 7-10 levels will be removed?

I suppose it doesn't, so using max here.

File: src/renderers/webgl/WebGLTextures.js (hunk 2 above)
Comment (mrdoob) [reply src/renderers/webgl/WebGLTextures.js]: It's not guaranteed, but lets assume that.

File: src/renderers/webgl/WebGLTextures.js (hunk 3 above)
Comment (mrdoob) [reply src/renderers/webgl/WebGLTextures.js]: @kenrussell do you know the answer for this?

**Comment by mrdoob** [comment]
Merging for now.

**Comment by mrdoob** [comment]
It'd be nice to compare how different is this from `PMREMGenerator`.

But, for people not using it, rough reflections look much better indeed!

screen shot 2018-03-06 at 7 34 29 pm

File: src/renderers/webgl/WebGLTextures.js (hunk 3 above)
Comment (kenrussell) [reply src/renderers/webgl/WebGLTextures.js]: Please see section 3.8.10.5 "Manual Mipmap Generation" of the OpenGL ES 3.0.5 spec.

It says that levels <code>level<sub>base+1</sub></code> through <code>q</code> are derived from the contents of <code>level<sub>base</sub></code>. Other mip levels are untouched.

I'm not sure what the intent of this code is, but probably you need to zero out all levels of the texture that were previously set, and set the max mip level to the one that will be set by GenerateMipmap.

File: src/renderers/webgl/WebGLTextures.js (hunk 2 above)
Comment (takahirox) [reply src/renderers/webgl/WebGLTextures.js]: OK. I wanna add comment about our assumption here later.

File: src/renderers/webgl/WebGLTextures.js (hunk 3 above)
Comment (mrdoob) [reply src/renderers/webgl/WebGLTextures.js]: @kenrussell thanks!

Comment (takahirox) [reply src/renderers/webgl/WebGLTextures.js]: I think I had just misunderstood a bit.
In the example I mentioned, 7-10 levels makes no sense (seems like we can set 1x1 image tho)
I can remove `max`.

**Comment by mrdoob** [comment]
In ChromeOS seems like it gets blurred too quickly?

screenshot 2018-03-07 at 22 54 39

**Comment by mrdoob** [comment]
On windows...

I see the same behavior with Microsoft Edge: 

edge

Chrome seems fine:

chrome

And so seems Firefox:

firefox

**Comment by takahirox** [comment]
Confirmed it with Windows + Edge here, too.

Hm, I guess it could be from the WebGL (extensions) capability?

**Comment by mrdoob** [comment]
Which extension would it be?

**Comment by takahirox** [comment]
I supposed float texture but Edge has. Hm, I might be wrong?

Chrome
image

Edge
image

**Comment by mrdoob** [comment]
I was not aware this code relies on floating point textures...?

**Comment by mrdoob** [comment]
I guess `maxMipLevel` that gets computed is different in Edge/Chrome OS?

**Comment by takahirox** [comment]
> I was not aware this code relied in floating point textures...?

Ah, sorry. I just said floating texture without thinking deeply because I haven't looked into the related shader code yet.

**Comment by takahirox** [comment]
> I guess maxMipLevel that gets computed is different in Edge/Chrome OS?

You meant, you guess the behavior of gl.generateMipmap() could be different?

**Comment by takahirox** [comment]
This issue seems appearing even in r90. I'm really not sure if it's related to mipmap yet.

**Comment by mrdoob** [comment]
> This issue seems appearing even in r90. I'm really not sure if it's related to mipmap yet.

That is a good point.

**Comment by mrdoob** [comment]
Wild guess... Maybe this is related to `window.devicePixelRatio`?

**Comment by mrdoob** [comment]
@takahirox is `window.devicePixelRatio` the same value in Edge and Chrome in your setup?

**Comment by WestLangley** [comment]
Another related point:

Our `material.roughness` is intended to be "perceptual" roughness. I would not be surprised if metallic and non-metallic materials responded differently to that property value. In any event, this is going to be somewhat subjective.

How the shader selects the mipmap level based on the `roughness` property is based on code copied from another source. That code can -- and may have to -- be changed.

**Comment by takahirox** [comment]
> is window.devicePixelRatio the same value in Edge and Chrome in your setup?

Same.

**Pull request merged**